File: 572-SubtreeOfAnotherTree/572-SubtreeOfAnotherTree.java
// Last updated: 11/2/2025, 8:22:27 PM
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
     public boolean isSubtree(TreeNode root, TreeNode subRoot) {
        if (root == null) return false;
        if (isSame(root, subRoot)) return true;
        return isSubtree(root.left, subRoot) || isSubtree(root.right, subRoot);
    }
    
    private boolean isSame(TreeNode root, TreeNode subRoot) {
        //first condition 
        if (root == null && subRoot == null) return true;
        
        //second condition
        // equivalent of (s == null && t != null) || (s != null) && (t==null)
        if (root == null || subRoot == null) return false;
        
        //third conition
        if (root.val != subRoot.val) return false;
        
        return isSame(root.left, subRoot.left) && isSame(root.right, subRoot.right);
    }
}

File: 572-SubtreeOfAnotherTree/572-SubtreeOfAnotherTree.py
# Last updated: 11/2/2025, 8:22:26 PM
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:

    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        if not root:
            return False
        if not subRoot:
            return False

        if self.isSameTree(root,subRoot):
            return True 

        return self.isSubtree(root.left, subRoot) or self.isSubtree(root.right,subRoot)

    def isSameTree(self, root,subRoot):
        if root == None and subRoot == None:
            return True  

        if root == None or subRoot == None:
            return False
        
        if root.val != subRoot.val:
            return False

        return self.isSameTree(root.left,subRoot.left) and self.isSameTree(root.right,subRoot.right)


        